rpki-validator/src/main/java/net/ripe/rpki/validator3/api/roas/ExportsController.java

```diff
@@ -123,74 +121,114 @@ public void exportCsv(HttpServletResponse response) throws IOException {
-                        Instant.ofEpochMilli(r.getNotBefore()).toString(),
-                        Instant.ofEpochMilli(r.getNotAfter()).toString(),
-                        r.getSerialNumber().toString()
-                ))
-                .distinct();
-
-        return new JsonExportExtended(validatedPrefixes);
+        try (CSVWriter writer = new CSVWriter(response.getWriter())) {
+            writer.writeNext(new String[]{"ASN", "IP Prefix", "Max Length", "Trust Anchor",
+                                          "Not Valid Before", "Not Valid After", "Serial Number"});
+            vrpExtendedStream().forEach(vrp ->
+                writer.writeNext(new String[]{
+                    vrp.getAsn(),
+                    vrp.getPrefix(),
+                    String.valueOf(vrp.getMaxLength()),
+                    vrp.getTa(),
+                    vrp.getNotBefore(),
+                    vrp.getNotAfter(),
+                    vrp.getSerialNumber()
+                }));
+        }
     }
 
-    @Value
-    private static class CsvRoaPrefix {
-        private String asn;
-        private String prefix;
-        private int maxLength;
-        private String trustAnchorName;
+    /**
+     * Get the stream of VRPs based only on ROAs, but having more fields.
+     */
+    private Stream<ExtendedRoa> vrpExtendedStream() {
+        return validatedRpkiObjects
+            .findCurrentlyValidatedRoaPrefixes()
+            .getObjects()
+            .map(r -> new ExtendedRoa(
+                String.valueOf(r.getAsn()),
+                r.getPrefix().toString(),
+                r.getEffectiveLength(),
+                r.getTrustAnchor().getName(),
+                Instant.ofEpochMilli(r.getNotBefore()).toString(),
+                Instant.ofEpochMilli(r.getNotAfter()).toString(),
+                r.getSerialNumber().toString()
+            ))
+            .distinct();
     }
 
+    /**
+     * Get the stream of VRPs based on ROAs - whitelist + SLURM assertions.
+     */
+    private Stream<VRP> vrpStream() {
+        Stream<VRP> validatedPrefixes = validatedRpkiObjects
+            .findCurrentlyValidatedRoaPrefixes()
+            .getObjects()
+            .filter(new IgnoreFiltersPredicate(ignoreFilters.all()).negate())
+            .map(r -> new VRP(
+                String.valueOf(r.getAsn()),
+                r.getPrefix().toString(),
+                r.getEffectiveLength(),
+                r.getTrustAnchor().getName())
+            );
+
+        final Stream<VRP> assertions = roaPrefixAssertions
+            .all()
+            .map(assertion -> new VRP(
+                assertion.getAsn().toString(),
+                assertion.getPrefix().toString(),
+                assertion.getMaxPrefixLength() != null ? assertion.getMaxPrefixLength() : assertion.getPrefix().getPrefixLength(),
+                "SLURM"
+            ));
+
+        return Stream.concat(validatedPrefixes, assertions).distinct();
```

Comment:
VRPs with an differing `ta` will not be seen as distinct (so SLURM versus from any of the trust anchors will both be there)

Reply:
I guess the point was to just return all VRPs regardless of their priority or origin, so conflict resolution and prioritising is up to the consumer.